HPCC-20641 Fix core in Esdl2Transformer

afishbeck opened the pull request:
Use ReadWriteLocks to protect lazy indexing of ESDL definition
types and methods.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
@timothyklemm @gaojx please review
@gaojx please include this commit in the build you use for testing

timothyklemm commented on esp/esdllib/esdl_transformer2.cpp:
```diff
@@ -1468,9 +1468,24 @@ void Esdl2Transformer::addMethod(Esdl2Base *item)
     }
 }
 
+void Esdl2Transformer::setMethod(const char *name, Esdl2Method *method)
```
Name is inconsistent with readMethodInfo and queryMethodInfo. Also inconsistent with addType relative to readType and queryType. Should it be addMethodInfo?

afishbeck replied:
I'll change to setMethodInfo() set because it is setting the value saved with name to be equal to this methodInfo.  addType just takes a type and doesn't directly set the name value pair.. although it does it internally by calling a set method.

afishbeck replied:
set in this case means a one to one relationship.  name=value.  In my general way of thinking, "append" or "add" would be a one to many or many to many relationship.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 54d77af5b70671983f8bf4cffb5f58b20bacb7dd
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |729 |1 |
|test (roxie) |879 |879 |0 |


[Errors:](https://gist.github.com/689f3ed84aa24a73f0abc38b33558756)
- test
  - thor
    - 657\. [superkey1](https://gist.github.com/HPCCSmoketest/689f3ed84aa24a73f0abc38b33558756#file-superkey1-w20181024-164702-1-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   42 sec (00:00:42) |  187 sec (00:03:07) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1313 sec (00:21:53) |   18 sec (00:00:18) | 1581 sec (00:26:21) |

AttilaVamos commented:
@afishbeck The Superkey1 error is not related to this PR.

afishbeck committed:
HPCC-20641 Fix core in Esdl2Transformer

Use ReadWriteLocks to protect lazy indexing of ESDL definition
types and methods.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

AttilaVamos commented:
@afishbeck  The Superkey1 error is not related to this PR.

afishbeck commented:
@timothyklemm @gaojx updated please review

HPCCSmoketest commented:
Process of PR-11868, label: desdlTxCore2 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 10937ef7afdf716037ce6b163c375f3a2fabed88
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |729 |1 |
|test (roxie) |879 |879 |0 |


[Errors:](https://gist.github.com/ec96235dd8258c62dbb768b60385fd40)
- test
  - thor
    - 657\. [superkey1](https://gist.github.com/HPCCSmoketest/ec96235dd8258c62dbb768b60385fd40#file-superkey1-w20181024-180921-1-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  188 sec (00:03:08) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1341 sec (00:22:21) |   20 sec (00:00:20) | 1588 sec (00:26:28) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 10937ef7afdf716037ce6b163c375f3a2fabed88
Build: success
Install hpccsystems-platform-community_7.0.0-rc3.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   43 sec (00:00:43) |  184 sec (00:03:04) |   95 sec (00:01:35) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1512 sec (00:25:12) |   18 sec (00:00:18) | 1914 sec (00:31:54) |

HPCCSmoketest commented:
Process of PR-11868, label: desdlTxCore2 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 10937ef7afdf716037ce6b163c375f3a2fabed88
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |729 |1 |
|test (roxie) |879 |879 |0 |


[Errors:](https://gist.github.com/1135325eefb671999eaa9e11a8bd00e5)
- test
  - thor
    - 657\. [superkey1](https://gist.github.com/HPCCSmoketest/1135325eefb671999eaa9e11a8bd00e5#file-superkey1-w20181025-055554-1-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  186 sec (00:03:06) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1321 sec (00:22:01) |   19 sec (00:00:19) | 1581 sec (00:26:21) |

HPCCSmoketest commented:
Process of PR-11868, label: desdlTxCore2 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 10937ef7afdf716037ce6b163c375f3a2fabed88
Build: success
Install hpccsystems-platform-community_7.0.0-rc3.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   37 sec (00:00:37) |  188 sec (00:03:08) |   95 sec (00:01:35) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1687 sec (00:28:07) |   19 sec (00:00:19) | 2088 sec (00:34:48) |

richardkchapman commented:
Should this target 7.0.2?
